geomesa-compute/src/main/scala/org/locationtech/geomesa/compute/spark/GeoMesaSpark.scala

```diff
@@ -187,6 +188,224 @@ object GeoMesaSpark extends LazyLogging {
     groupedByDay.map { case (date, iter) => (date, iter.size) }
   }
 
+  def shallowJoin(sc: SparkContext, coveringSet: RDD[SimpleFeature], data: RDD[SimpleFeature], key: String): RDD[SimpleFeature] = {
+    // Broadcast sfts to executors
+    val sfts = sc.broadcast(GeoMesaSparkKryoRegistrator.typeCache.values.map{ sft =>
+      (sft.getTypeName, SimpleFeatureTypes.encodeType(sft))
+    }.toArray)
+
+    data.foreachPartition{ iter =>
+      sfts.value.foreach{ case (name, spec) =>
+        val sft = SimpleFeatureTypes.createType(name, spec)
+        GeoMesaSparkKryoRegistrator.putType(sft)
+      }
+    }
+
+    // Broadcast covering set to executors
+    val broadcastedCover = sc.broadcast(coveringSet.collect)
+
+    // Key data by cover name
+    val keyedData = data.mapPartitions { iter =>
+      import org.locationtech.geomesa.utils.geotools.Conversions._
+
+      iter.flatMap { sf =>
+        // Iterate over covers until a match is found
+        val it = broadcastedCover.value.iterator
+        var container = ""
+        while (it.hasNext && container == "") {
+          val cover = it.next()
+          // If the cover's polygon contains the feature, set the container
+          if (cover.geometry.contains(sf.geometry)) {
+            container = cover.getAttribute(key).asInstanceOf[String]
+          }
+        }
+        // return the found cover as the key
+        if (container != "") {
+          Some(container, sf)
+        } else {
+          None
+        }
+      }
+    }
+
+    // Get the indices and types of the attributes that can be aggregated and send them to the partitions
+    val featureAttributes = data.first.getAttributes.toSeq
+    val countableIndices = featureAttributes.toIndexedSeq.indices.flatMap( { index =>
+      // Skip the id
+      if (index != 0) {
+        featureAttributes(index) match {
+          case attr if attr != null && attr.getClass == classOf[Integer] =>
+            Some(index, "Integer")
+          case attr if attr != null && attr.getClass == classOf[java.lang.Long] =>
+            Some(index, "Long")
+          case attr if attr != null && attr.getClass == classOf[java.lang.Double] =>
```

Comment:
null objects wouldn't pattern match to a type - it would be caught by your `case _ =>`
```
scala> val i: Integer = null
i: Integer = null

scala> i match { case m: Integer => println("matched integer"); case _ => println("matched null") }
matched null
```